RandLAPACK/gpu_functions/rl_cuda_kernels.cuh

```diff
@@ -0,0 +1,934 @@
+) {
+    if (copy_upper_triangle) {
+        // Only copying the upper-triangular portion of the original
+        for (int64_t i = threadIdx.x + blockDim.x * blockIdx.x; i < n; i += blockDim.x * gridDim.x) {
+            for (int64_t j = threadIdx.y + blockDim.y * blockIdx.y; j <= i; j += blockDim.y * gridDim.y) {
+                A_cpy[(ldac * i) + j] = A[(lda * i) + j];
+            }
+        }
+    } else {
+        for (int64_t i = threadIdx.x + blockDim.x * blockIdx.x; i < n; i += blockDim.x * gridDim.x) {
+            for (int64_t j = threadIdx.y + blockDim.y * blockIdx.y; j < m; j += blockDim.y * gridDim.y) {
+                A_cpy[(ldac * i) + j] = A[(lda * i) + j];
+            }
+        }
+    }
+}
+#endif
+
+template <typename T>
+void naive_rank_est(
+    cudaStream_t stream, 
+    int64_t n,
+    const T alpha,
+    const T* A,
+    int64_t lda,
+    int64_t& rank
+) {
+#ifdef USE_CUDA
+    int64_t* rank_device;
+    cudaMalloc(&rank_device, sizeof(int64_t));
+
+    constexpr int threadsPerBlock{128};
+    int64_t num_blocks_ger{(n + threadsPerBlock - 1) / threadsPerBlock};
+    naive_rank_est<<<num_blocks_ger, threadsPerBlock, sizeof(int64_t), stream>>>(n, alpha, A, lda, rank_device);
+#endif
+    // It seems that these synchs are required, after all
+    cudaStreamSynchronize(stream);
+    cudaMemcpy(&rank, rank_device, sizeof(int64_t), cudaMemcpyDeviceToHost);
+    cudaStreamSynchronize(stream);
+    cudaFree(rank_device);
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch naive_rank_est. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+template <typename T>
+void all_of(
+    cudaStream_t stream, 
+    int64_t n,
+    const T alpha,
+    const T* A,
+    bool& all_equal
+) {
+#ifdef USE_CUDA
+    bool* all_equal_device;
+    cudaMalloc(&all_equal_device, sizeof(bool));
+
+    constexpr int threadsPerBlock{128};
+    int64_t num_blocks_ger{(n + threadsPerBlock - 1) / threadsPerBlock};
+    all_of<<<num_blocks_ger, threadsPerBlock, sizeof(bool), stream>>>(n, alpha, A, all_equal_device);
+    cudaStreamSynchronize(stream);
+    cudaMemcpy(&all_equal, all_equal_device, sizeof(bool), cudaMemcpyDeviceToHost);
+    cudaStreamSynchronize(stream);
+    cudaFree(all_equal_device);
+#endif
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch all_equal. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+template <typename T>
+void ger_gpu(
+    cudaStream_t stream, 
+    int64_t m, 
+    int64_t n, 
+    T alpha, 
+    T *x, 
+    int64_t incx, 
+    T *y, 
+    int64_t incy, 
+    T* A, 
+    int64_t lda
+) {
+
+#ifdef USE_CUDA
+    constexpr int threadsPerBlock{128};
+    int64_t num_blocks_ger{(m + threadsPerBlock - 1) / threadsPerBlock};
+    ger_gpu<<<num_blocks_ger, threadsPerBlock, 0, stream>>>(m, n, alpha, x, incx, y, incy, A, lda);
+#endif
```

Comment:
Similar comment as before. It seems dangerous to have the USE_CUDA in here. We want to raise errors if someone calls a function that they shouldn't.

Reply:
Fixed.